docs/2024/puzzles/day21.md

```diff
@@ -2,10 +2,243 @@ import Solver from "../../../../../website/src/components/Solver.js"
 
 # Day 21: Keypad Conundrum
 
+by [@mbovel](https://github.com/mbovel)
+
 ## Puzzle description
 
 https://adventofcode.com/2024/day/21
 
+## Data structures
+
+We begin by defining the data structures used to represent the keypads and positions on the grid. We start with a `Pos` case class for 2D coordinates. Then, we define two keypads—`numericalKeypad` and `directionalKeypad`—as `Map[Char, Pos]`, with their corresponding sets of valid positions as `Set[Pos]`.
+
+```scala mdoc:silent
+case class Pos(x: Int, y: Int):
+  def +(other: Pos) = Pos(x + other.x, y + other.y)
+  def -(other: Pos) = Pos(x - other.x, y - other.y)
+  def projX = Pos(x, 0)
+  def projY = Pos(0, y)
+
+val numericalKeypad = Map(
+  '7' -> Pos(0, 0), '8' -> Pos(1, 0), '9' -> Pos(2, 0),
+  '4' -> Pos(0, 1), '5' -> Pos(1, 1), '6' -> Pos(2, 1),
+  '1' -> Pos(0, 2), '2' -> Pos(1, 2), '3' -> Pos(2, 2),
+                    '0' -> Pos(1, 3), 'A' -> Pos(2, 3),
+)
+val numericalKeypadPositions = numericalKeypad.values.toSet
```

Comment:
I'm really nitpicking here, but it ought to be "numeric" throughout rather than "numerical"